### DISCUSSION TITLE
When messages are consumed normally, backlogsize does not decrease, but increases. The backlog quota exceeded

### DISCUSSION THREAD

**Discussion by imitateinfant** [apache/pulsar#22214] [Q&A]
Hi, everyone. I have a problem and I need help
I use the java-api to create producers and consumers. As shown below, the production rate and consumption rate are consistent and should not result in a message backlog, but the backlog keeps increasing and eventually exceeds the rated quota.
{
  "msgRateIn" : 29.516710605242075,
  "msgThroughputIn" : 94924.34130437447,
  "msgRateOut" : 29.533377153731685,
  "msgThroughputOut" : 94983.67425016942,
  "averageMsgSize" : 3215.9525691699605,
  "storageSize" : 611400548,
  "backlogSize" : 586682372,
  "publishers" : [ {
    "msgRateIn" : 29.516710605242075,
    "msgThroughputIn" : 94924.34130437447,
    "averageMsgSize" : 3215.0,
    "producerId" : 0,
    "metadata" : { },
    "address" : "/172.16.253.15:44900",
    "producerName" : "ZHCLOUD_SCHEDULE_MQTT0111",
    "connectedSince" : "2024-03-07T09:12:28.127+08:00",
    "clientVersion" : "2.6.1"
  } ],
  "subscriptions" : {
    "ZHCLOUD_SCHEDULE_MQTT01" : {
      "msgRateOut" : 29.533377153731685,
      "msgThroughputOut" : 94983.67425016942,
      "msgRateRedeliver" : 0.0,
      "msgBacklog" : 4,
      "blockedSubscriptionOnUnackedMsgs" : false,
      "msgDelayed" : 0,
      "unackedMessages" : 4,
      "type" : "Shared",
      "msgRateExpired" : 0.0,
      "lastExpireTimestamp" : 0,
      "consumers" : [ {
        "msgRateOut" : 5.966675555720466,
        "msgThroughputOut" : 18555.944311003237,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-d0083550e73049fd93abe70eca0ed274",
        "availablePermits" : 666,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60194",
        "connectedSince" : "2024-03-07T09:12:28.268+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 5.900008761709678,
        "msgThroughputOut" : 18585.24426637391,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-d5024930e4974f9d94ccd02bb83e3737",
        "availablePermits" : 981,
        "unackedMessages" : 1,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60200",
        "connectedSince" : "2024-03-07T09:12:28.269+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 5.766675244403871,
        "msgThroughputOut" : 18367.52732108464,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-99a64e55d9c84c06ac02d9cd67aba399",
        "availablePermits" : 623,
        "unackedMessages" : 2,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60196",
        "connectedSince" : "2024-03-07T09:12:28.278+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 5.966675505401427,
        "msgThroughputOut" : 19061.528236794853,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-3d15d386a3fd4650ab25a3b609cd0d33",
        "availablePermits" : 800,
        "unackedMessages" : 1,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60198",
        "connectedSince" : "2024-03-07T09:12:28.289+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 5.933342086496246,
        "msgThroughputOut" : 20413.43011491278,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-211f5fd8c8d4466ca0e0d82dba203f11",
        "availablePermits" : 741,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60202",
        "connectedSince" : "2024-03-07T09:12:28.29+08:00",
        "clientVersion" : "2.5.0"
      } ],
      "isReplicated" : false
    }
  },
  "replication" : { },
  "deduplicationStatus" : "Disabled",
  "bytesInCounter" : 586477073,
  "msgInCounter" : 182070
}
When I shut down the producer program, there was a lot of backlogsize, but the consumer was not spending
{
  "msgRateIn" : 0.0,
  "msgThroughputIn" : 0.0,
  "msgRateOut" : 0.0,
  "msgThroughputOut" : 0.0,
  "averageMsgSize" : 0.0,
  "storageSize" : 746435354,
  "backlogSize" : 721717178,
  "publishers" : [ ],
  "subscriptions" : {
    "ZHCLOUD_SCHEDULE_MQTT01" : {
      "msgRateOut" : 0.0,
      "msgThroughputOut" : 0.0,
      "msgRateRedeliver" : 0.0,
      "msgBacklog" : 1,
      "blockedSubscriptionOnUnackedMsgs" : false,
      "msgDelayed" : 0,
      "unackedMessages" : 1,
      "type" : "Shared",
      "msgRateExpired" : 0.0,
      "lastExpireTimestamp" : 0,
      "consumers" : [ {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-d0083550e73049fd93abe70eca0ed274",
        "availablePermits" : 809,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60194",
        "connectedSince" : "2024-03-07T09:12:28.268+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-d5024930e4974f9d94ccd02bb83e3737",
        "availablePermits" : 614,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60200",
        "connectedSince" : "2024-03-07T09:12:28.269+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-99a64e55d9c84c06ac02d9cd67aba399",
        "availablePermits" : 821,
        "unackedMessages" : 1,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60196",
        "connectedSince" : "2024-03-07T09:12:28.278+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-3d15d386a3fd4650ab25a3b609cd0d33",
        "availablePermits" : 972,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60198",
        "connectedSince" : "2024-03-07T09:12:28.289+08:00",
        "clientVersion" : "2.5.0"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-211f5fd8c8d4466ca0e0d82dba203f11",
        "availablePermits" : 754,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "address" : "/172.16.253.24:60202",
        "connectedSince" : "2024-03-07T09:12:28.29+08:00",
        "clientVersion" : "2.5.0"
      } ],
      "isReplicated" : false
    }
  },
  "replication" : { },
  "deduplicationStatus" : "Disabled",
  "bytesInCounter" : 721511879,
  "msgInCounter" : 223911
}
When I restart the consumer app, the backlogsize will disappear
{
  "msgRateIn" : 0.0,
  "msgThroughputIn" : 0.0,
  "msgRateOut" : 0.0,
  "msgThroughputOut" : 0.0,
  "averageMsgSize" : 0.0,
  "storageSize" : 75088016,
  "backlogSize" : 7964,
  "publishers" : [ ],
  "subscriptions" : {
    "ZHCLOUD_SCHEDULE_MQTT01" : {
      "msgRateOut" : 0.0,
      "msgThroughputOut" : 0.0,
      "msgRateRedeliver" : 0.0,
      "msgBacklog" : 0,
      "blockedSubscriptionOnUnackedMsgs" : false,
      "msgDelayed" : 0,
      "unackedMessages" : 0,
      "type" : "Shared",
      "msgRateExpired" : 0.0,
      "lastExpireTimestamp" : 0,
      "consumers" : [ {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-0e74a69ad30b41a9994255eeada548c7",
        "availablePermits" : 1000,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "connectedSince" : "2024-03-07T11:23:05.087+08:00",
        "clientVersion" : "2.5.0",
        "address" : "/172.16.253.24:37222"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-796d10a7c8c942188b1cd8e943a21381",
        "availablePermits" : 1000,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "connectedSince" : "2024-03-07T11:23:05.087+08:00",
        "clientVersion" : "2.5.0",
        "address" : "/172.16.253.24:37216"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-20ecbb92007d42e3987d0ea97ac07618",
        "availablePermits" : 999,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "connectedSince" : "2024-03-07T11:23:05.088+08:00",
        "clientVersion" : "2.5.0",
        "address" : "/172.16.253.24:37218"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-a48fb95a267049c1b813ab81ff1fea41",
        "availablePermits" : 1000,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "connectedSince" : "2024-03-07T11:23:05.092+08:00",
        "clientVersion" : "2.5.0",
        "address" : "/172.16.253.24:37220"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "ZHCLOUD_SCHEDULE_MQTT01-04ea37a71c0d4bbab6b25dd133fa6896",
        "availablePermits" : 1000,
        "unackedMessages" : 0,
        "blockedConsumerOnUnackedMsgs" : false,
        "metadata" : { },
        "connectedSince" : "2024-03-07T11:23:05.105+08:00",
        "clientVersion" : "2.5.0",
        "address" : "/172.16.253.24:37226"
      } ],
      "isReplicated" : false
    }
  },
  "replication" : { },
  "deduplicationStatus" : "Disabled",
  "bytesInCounter" : 721511879,
  "msgInCounter" : 223911
}

**Comment by lhotari** [comment]
You seem to be using Pulsar 2.5.0 and 2.6.1 clients. A lot of bugs have been fixed since then. I'd recommend upgrading to Pulsar 3.0.2 client which is the LTS version client. 
How about your broker? Which version is it? I'd also recommend 3.0.3 for that.
Support has ended for all other than Pulsar 3.0.x and 3.2.x:
https://pulsar.apache.org/contribute/release-policy/#supported-versions

**Reply by imitateinfant** [reply]
Yes, both my client and broker are using version 2.5.0, I will try to upgrade to the recommended version and see if that works.thank you
